docs/proposals/pod-security-context.md

```diff
@@ -0,0 +1,154 @@
+
+1.  Allow users to cleanly express security settings that are applicable to the entire pod
+2.  Allow users to express default settings for the container-level context and override only the
+    settings that need to be differentiated from the pod security context in containers where
+    required
+
+Goals of this design:
+
+1.  Describe the use-cases for which a pod-level security context is necessary
+2.  Describe the model changes and associated implementation changes for the refactor
+
+## Use Cases
+
+1.  As a user, I want to be able to specify a default security context for all containers in a pod
+2.  As a user, I want to be able to override applicable pieces of the pod security context for
+    individual containers as needed
+
+### Use Case: Default security context for all containers
+
+By default, all containers in a pod should share the same security settings, and those settings
+should only have to be specified once.  If users are forced to specify the same settings for each
+container, it requires more complex validations and allows for transcription errors between
+containers
+
+Additionally, use-cases for sharing resources such as volumes are much easier to implement if a user
+can describe a pod-level security context.
+
+### Use Case: Override pod security context for container
+
+Some use-cases require the containers in a pod to run with different security settings.  As an
+example, a user may want to have a pod with two containers, one of which runs as root with the
+privileged setting, and one that runs as a non-root UID.  To support use-cases like this, it should
+be possible to override appropriate (ie, not intrinsically pod-level) security settings for
+individual containers.
+
+## Proposed Design
+
+The `SecurityContext` type should be replaced by two new, similar types: `ContainerSecurityContext`
+and `PodSecurityContext`:
+
+### ContainerSecurityContext
+
+The `ContainerSecurityContext` type models the security settings that specified for a container:
+
+```go
+type ContainerSecurityContext struct {
+    Capabilities              *Capabilities
+    Privileged                *bool
+    SELinuxOptions            *SELinuxOptions
+    RunAsUser                 *int64
+    RunAsNonRoot              bool
+    RunAsGroup                *int64
+    RunWithSupplementalGroups []int64
+}
+```
+
+The `ContainerSecurityContext` type is very similar to the existing `SecurityContext` type, with
+two additions:
+
+1.  The `RunAsGroup` field specifies the GID the container process runs as
+2.  The `RunWithSupplementalGroups` field specifies additional groups the container process should
+    be in
+
+The addition of these fields enables scenarios where containers share resources via groups.
+
+### PodSecurityContext
+
+The `PodSecurityContext` type expresses pod-level security settings and the default container-level
+settings:
+
+```go
+type PodSecurityContext struct {
+    HostNetwork       bool
+    HostIpc           bool
```

Comment:
Containers may not want different MCS labels but they may want different
UIDs.  For backwards compatibility you'll probably still have to preserve
the label on the container, but I would agree label differentiation has no
concrete use case.

On Wed, Aug 26, 2015 at 4:25 PM, Paul Morie <notifications@github.com>
wrote:

> In docs/proposals/pod-security-context.md
> <https://github.com/kubernetes/kubernetes/pull/12823#discussion_r38023832>
> :
>
> > +
> > +1.  The `RunAsGroup` field specifies the GID the container process runs as
> > +2.  The `RunWithSupplementalGroups` field specifies additional groups the container process should
> > +    be in
> > +
> > +The addition of these fields enables scenarios where containers share resources via groups.
> > +
> > +### PodSecurityContext
> > +
> > +The `PodSecurityContext` type expresses pod-level security settings and the default container-level
> > +settings:
> > +
> > +```go
> > +type PodSecurityContext struct {
> > +    HostNetwork       bool
> > +    HostIpc           bool
>
> @bgrant0607 <https://github.com/bgrant0607> I would say the salient
> pieces for volumes are probably:
>
>    1. Making UID/GID settable on a pod level
>    2. Adding support for pod level supplemental groups
>    3. Making SELinux context exclusively pod level (maybe)
>
> @smarterclayton <https://github.com/smarterclayton> I am wondering
> whether actually have any use-cases for differentiating the security
> context of two containers in the same pod. For example, if two containers
> have different MCS labels, there is no way to share a volume between them
> without relaxing the SELinux context so that any container would be able to
> access it (from an SELinux standpoint). Do you know whether it's actually a
> requirement to be able to differentiate the SELinux contexts of two
> containers in the same pod?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12823/files#r38023832>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift